contracts/Controller.sol

```diff
@@ -124,38 +130,89 @@ contract Controller is Ownable {
      */
-    // function getVaultBalances(address _owner, uint256 _vaultId) external view returns (MarginAccount.Vault memory) {
-    // }
+    function getVaultBalances(address _owner, uint256 _vaultId) external view returns (MarginAccount.Vault memory) {
+        MarginAccount.Vault memory vault = getVault(_owner, _vaultId);
+
+        if ((vault.shortOtokens.length != 0) && (isExpired(vault.shortOtokens[0]))) {
+            address calculatorModule = AddressBookInterface(addressBook).getMarginCalculator();
+            MarginCalculatorInterface calculator = MarginCalculatorInterface(calculatorModule);
+
+            (uint256 netValue, ) = calculator.getExcessMargin(vault, vault.shortOtokens[0]);
+            vault.collateralAmounts[0] = netValue;
+
+            return vault;
+        }
+
+        return vault;
+    }
 
     /**
      * @dev return if an expired oToken contract’s price has been finalized. Returns true if the contract has expired AND the oraclePrice at the expiry timestamp has been finalized.
      * @param _otoken The address of the relevant oToken.
      * @return A boolean which is true if and only if the price is finalized.
      */
-    // function isPriceFinalized(address _otoken) external view returns (bool) {
-    // }
+    function isPriceFinalized(address _otoken) external view returns (bool) {
+        address oracleModule = AddressBookInterface(addressBook).getOracle();
+        OracleInterface oracle = OracleInterface(oracleModule);
+
+        OtokenInterface otoken = OtokenInterface(_otoken);
+
+        address otokenUnderlyingAsset = otoken.underlyingAsset();
+        address otokenStrikeAsset = otoken.strikeAsset();
+        address otokenCollateralAsset = otoken.collateralAsset();
+        uint256 otokenExpiryTimestamp = otoken.expiryTimestamp();
+
+        bytes32 batch = keccak256(
+            abi.encode(otokenUnderlyingAsset, otokenStrikeAsset, otokenCollateralAsset, otokenExpiryTimestamp)
+        );
+
+        return oracle.isDisputePeriodOver(batch, otokenExpiryTimestamp);
+    }
 
     /**
-     * @notice Return a specific vault.
-     * @param _owner vault owner.
-     * @param _vaultId vault.
-     * @return Vault struct
+     * @notice get batch for a specific otoken address
+     * @dev batch is the hash of option underlying, strike, collateral assets and the expiry timestamp
+     * @param _otoken otoken address
+     * @return batch hash in bytes32
      */
-    // function getVault(address _owner, uint256 _vaultId) public view returns (MarginAccount.Vault memory) {
-    // }
+    function getBatch(address _otoken) external view returns (bytes32) {
+        OtokenInterface otoken = OtokenInterface(_otoken);
+
+        return
+            keccak256(
+                abi.encode(
+                    otoken.underlyingAsset(),
+                    otoken.strikeAsset(),
+                    otoken.collateralAsset(),
+                    otoken.expiryTimestamp()
+                )
+            );
+    }
```

Comment:
I think we should have this helper function in another contract like `OUtil.sol` , and both `Controller` and `Calculator` can inherit it.

Reply:
Where does calculator need it ?

Reply:
It needs to get the batch price too. (for getExpiredCashValue)

Reply:
Will move it in another PR once Calculator get merged.